networkit/cpp/graph/GraphTools.cpp

```diff
@@ -200,15 +200,66 @@ double density(const Graph &G) noexcept {
     return m / (n * (n - 1));
 }
 
-double volume(const Graph &G) {
-    if (G.isDirected())
-        throw std::runtime_error("Volume computation is only supported for undirected graphs.");
+double inVolume(const Graph &G, const std::unordered_set<node> &nodes) {
+    double inVolume = 0.0;
+    bool directed = G.isDirected() ? false : true;
+
+    // Compute volume of subgraph
+    if (!nodes.empty() && G.numberOfNodes() != nodes.size()) {
+        for (auto it = std::begin(nodes); it != std::end(nodes); ++it) {
+            inVolume += G.weightedDegreeIn(*it, directed);
+
+            // This removes all incoming edges from the rest of the graph to the subgraph
+            G.forInNeighborsOf(*it, [&](node v) {
+                std::unordered_set<node>::const_iterator neighborIt = nodes.find(v);
+                if (neighborIt == nodes.end()) {
+                    inVolume -= static_cast<double>(G.weight(v, *it));
+                }
+            });
+        }
+        // Compute volume of the complete graph
+    } else {
+        if (G.isWeighted()) {
+            inVolume = G.totalEdgeWeight();
+        } else {
+            inVolume = G.numberOfEdges();
+        }
+        if (!G.isDirected()) {
+            inVolume *= 2.0;
+        }
+    }
+    return inVolume;
+}
 
-    if (!G.isWeighted()) {
-        return 2.0 * G.numberOfEdges();
+double outVolume(const Graph &G, const std::unordered_set<node> &nodes) {
+    double outVolume = 0.0;
+    bool directed = G.isDirected() ? false : true;
+
+    // Compute volume of subgraph
+    if (!nodes.empty() > 0 && G.numberOfNodes() != nodes.size()) {
+        for (auto it = std::begin(nodes); it != std::end(nodes); ++it) {
+            outVolume += G.weightedDegree(*it, directed);
+
+            // This removes all outgoing edges from the subgraph to the rest
+            G.forNeighborsOf(*it, [&](node v) {
+                std::unordered_set<node>::const_iterator neighborIt = nodes.find(v);
+                if (neighborIt == nodes.end()) {
+                    outVolume -= static_cast<double>(G.weight(*it, v));
+                }
```

Comment:
Same here: For weighted graphs, this iterates over the neighbors twice: Once to collect the edge weights in `weightedDegree` and once here to remove some of them again. Why don't we directly only add those edge weights to neighbors that are in the set, i.e., remove the call to `G.weightedDegree` and add weights in the second loop when `neighborIt != nodes.end()`.